pkg/substitution/substitution.go

```diff
@@ -87,93 +98,107 @@ func ValidateVariableProhibited(name, value, prefix, locationName, path string,
 func ValidateVariableIsolated(name, value, prefix, locationName, path string, vars sets.String) *apis.FieldError {
-	if vs, present, _ := extractVariablesFromString(value, prefix); present {
-		firstMatch, _ := extractExpressionFromString(value, prefix)
-		for _, v := range vs {
-			v = strings.TrimSuffix(v, "[*]")
-			if vars.Has(v) {
-				if len(value) != len(firstMatch) {
-					return &apis.FieldError{
-						Message: fmt.Sprintf("variable is not properly isolated in %q for %s %s", value, locationName, name),
-						Paths:   []string{path + "." + name},
-					}
+	vs, err := extractVariablesFromString(value, prefix)
+	if err != nil {
+		return &apis.FieldError{
+			Message: fmt.Sprintf("extractVariablesFromString failed: %v", err),
+			Paths:   []string{""},
+		}
+	}
+
+	firstMatch, _ := extractExpressionFromString(value, prefix)
+	for _, v := range vs {
+		v = strings.TrimSuffix(v, "[*]")
+		if vars.Has(v) {
+			if len(value) != len(firstMatch) {
+				return &apis.FieldError{
+					Message: fmt.Sprintf("variable is not properly isolated in %q for %s %s", value, locationName, name),
+					Paths:   []string{path + "." + name},
 				}
 			}
 		}
 	}
+
 	return nil
 }
 
 // ValidateVariableIsolatedP verifies that variables matching the relevant string expressions are completely isolated if present.
 func ValidateVariableIsolatedP(value, prefix string, vars sets.String) *apis.FieldError {
-	if vs, present, errString := extractVariablesFromString(value, prefix); present {
-		if errString != "" {
-			return &apis.FieldError{
-				Message: errString,
-				Paths:   []string{""},
-			}
-
+	vs, err := extractVariablesFromString(value, prefix)
+	if err != nil {
+		return &apis.FieldError{
+			Message: fmt.Sprintf("extractVariablesFromString failed: %v", err),
+			Paths:   []string{""},
 		}
-		firstMatch, _ := extractExpressionFromString(value, prefix)
-		for _, v := range vs {
-			v = strings.TrimSuffix(v, "[*]")
-			if vars.Has(v) {
-				if len(value) != len(firstMatch) {
-					return &apis.FieldError{
-						Message: fmt.Sprintf("variable is not properly isolated in %q", value),
-						// Empty path is required to make the `ViaField`, … work
-						Paths: []string{""},
-					}
+	}
+
+	firstMatch, _ := extractExpressionFromString(value, prefix)
+	for _, v := range vs {
+		v = strings.TrimSuffix(v, "[*]")
+		if vars.Has(v) {
+			if len(value) != len(firstMatch) {
+				return &apis.FieldError{
+					Message: fmt.Sprintf("variable is not properly isolated in %q", value),
+					// Empty path is required to make the `ViaField`, … work
+					Paths: []string{""},
 				}
 			}
 		}
 	}
+
 	return nil
 }
 
 // Extract a the first full string expressions found (e.g "$(input.params.foo)"). Return
-// "" and false if nothing is found.
-func extractExpressionFromString(s, prefix string) (string, bool) {
+// "" and nil if nothing is found.
+func extractExpressionFromString(s, prefix string) (string, error) {
```

Comment:
I think it would make sense to leave this function signature as is, and move the regexes into global variables instead of using Compile

Reply:
Since the regex string/pattern needs the input `prefix`, I guess it might not be possible to make it global variables?

Reply:
ah ok makes sense